### DISCUSSION TITLE
Code modification and application method

### DISCUSSION THREAD

**Discussion by DoyoonKim0209** [IBM/aihwkit#621] [Installation Q&A]
I want to apply the changes I made to the code in AIHWKIT and run the simulation.

I applied the method on Linux, but it doesn't seem to work well.

The method I used is as follows:

After downloading the files with git clone, I modified the cpp code and then executed the following code:

python setup.py build_ext --inplace -DUSE_CUDA=ON -DRPU_CUDA_ARCHITECTURES='86'

If I am wrong, I would like to know what method I should use.

**Comment by maljoras** [comment]
Hi @DoyoonKim0209 , that looks correct. If it doesn't work, can you post the error message? 

Note, that it is in general  somewhat more involved to modify C++ files, so you might first want to familiarize yourself with it. For most things it should not be necessary. All simulation settings are accessible from the python frontend.

**Reply by DoyoonKim0209** [reply]
There doesn't seem to be a specific error message. The installation runs smoothly, but there is a behavior that doesn't seem to reflect the changes made in the code. What I intended to do was to ensure that when the chopper changes in the src-rpucuda-rpu_chopped_transfer_device.cpp file, the sign of the chopper does not change, so that c-TTv2 behaves like TTv2, and AGAD changes only the moving average without changing the sign of the chopper. However, it didn't work out that way.

**Reply by DoyoonKim0209** [reply]
if (FROM_DEVICE_IDX == 0 && par.in_chop_prob > (T)0.0) {
      // randomize in_choppers
      if (par.in_chop_random) {
        if (this->rw_rng_.sampleUniform() < par.in_chop_prob) {
          in_chopper_[(size_t)(i_slice_start + i)] = !in_chopper_[(size_t)(i_slice_start + i)];
        }
      } else {
        const uint64_t n_samples = MAX((int)ceil((T)1.0 / par.in_chop_prob), 2);
        if ((this->transfer_counter_ % n_samples) + 1 == n_samples) {
          in_chopper_[(size_t)(i_slice_start + i)] = !in_chopper_[(size_t)(i_slice_start + i)];
        }
      }
    }
  }

I made modifications by either deleting this part or removing the exclamation mark, but both resulted in behavior similar to c-TTv2.

**Reply by maljoras-sony** [reply]
You can simply set `in_chop_prop=0`, see my comment below. However, also note that there are typically 2 implementations, one for CPU one optimized for CUDA. If you just change the CPU version, then the CUDA version is not affected, so that if you run it on GPU, the old version is simulated.

**Comment by maljoras** [comment]
Just as an advice, I would not try to modify the C++ files if it can be avoided. It is easy to introduce new errors if you are not carefully testing for it. Changing c-TTv2 to TTv2 is actually already supported without any changes you just have to set the chopper probability (`in_chop_prob`) to zero. For AGAD chopper flips are essential, because otherwise a leaky reference would remove the DC component.
